CARDS-2509: Reference question should support conditions based on souce form metadata

acrowthe opened the pull request:
Add optional condition properties to reference questions:
- conditionalProperty: A property on the referenced form to be evaluated
- conditionalOperator: The operator that should be used for the comparison. Supports all operators supported by Conditional Sections
- conditionalValue: The value that the property should be compared against
- conditionalFallback: The value that should be saved instead of the referenced value if the condition is not satisfied
- conditionalType: The type of the property that is being compared

Refactored backend conditional section code in order to reuse conditional evaluation code

Update Reference Test forms to add testing support for new conditional references
- Added new section with 4 new questions to the bottom of ReferenceTestCopied. These questions are conditional on ReferenceTestUser being completed
- Made one question in ReferenceTestUser mandatory to allow the form to be edited between incomplete and complete


To test:
- Verify that form completion status works with mandatory questions in conditional sections as before as the code that updates completion status was adjusted in order to be reusable
- Verify that the new questions in ReferenceTestCopied show:
  - Fallback values when ReferenceTestUser is incomplete
  - The values from ReferenceTestUser when that form is complete
- Make sure to test reference answers on brand new forms and on forms that have been edited

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
Add optional condition properties to reference questions:
- conditionalProperty: A property on the referenced form to be evaluated
- conditionalOperator: The operator that should be used for the comparison.
                       Supports all operators supported by Conditional Sections
- conditionalValue: The value that the property should be compared against
- conditionalFallback: The value that should be saved instead of the
                       referenced value if the condition is not satisfied
- conditionalType: The type of the property that is being compared

Refactored backend conditional section code in order to reuse conditional evaluation code

Update Reference Test forms to add testing support for new conditional references
- Added new section with 4 new questions to the bottom of ReferenceTestCopied.
  These questions are conditional on ReferenceTestUser being completed
- Made one question in ReferenceTestUser mandatory to allow the form to be edited
  between incomplete and complete

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
Add optional error message when a conditional reference is not satisfied
- Allows non-text questions to have a text based error display

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
- Add invalid source flag to reference questions with no source

sdumitriu commented on modules/data-model/forms/api/src/main/java/io/uhndata/cards/forms/api/FormUtils.java:
```diff
@@ -50,6 +50,9 @@ public interface FormUtils
     String RELATED_SUBJECTS_PROPERTY = "relatedSubjects";
 
+    /** The name of a property on an Answer, Section or Form node that holds the status flags for that node.*/
+    String STATUS_FLAGS = "statusFlags";
```
```suggestion
    String STATUS_FLAGS_PROPERTY = "statusFlags";
```

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersChangedListener.java:
```diff
@@ -58,8 +60,8 @@
-    /** Answer's property name. **/
-    public static final String VALUE = "value";
+    /** Property on an answer node that stores the a reference to the question. */
+    public static final String QUESTION = "question";
```
There's also `FormUtils.QUESTION_PROPERTY`

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersChangedListener.java:
```diff
@@ -191,31 +187,234 @@ private void checkAndUpdateAnswersValues(final NodeIterator nodeIterator, final
+     * @param referenceAnswer The reference answer to copy the value into
+     * @param sourceNode The source answer node that the value is being copied from
+     * @throws RepositoryException if an unexpected error occurs
+     */
```
Dangling comment?

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersChangedListener.java:
```diff
@@ -191,31 +187,234 @@ private void checkAndUpdateAnswersValues(final NodeIterator nodeIterator, final
+     */
+
+    /**
+     * Fill out a refernce answer with the value specified as the fallback value.
```
```suggestion
     * Fill out a reference answer with the value specified as the fallback value.
```

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersChangedListener.java:
```diff
@@ -191,31 +187,234 @@ private void checkAndUpdateAnswersValues(final NodeIterator nodeIterator, final
 
-    private boolean isNotSame(final Property source, final Node reference) throws RepositoryException
+    /**
+     * Fill out a refernce answer with a value copied from the referenced question.
```
```suggestion
     * Fill out a reference answer with a value copied from the referenced question.
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/utils/ConditionalUtils.java:
I want to remove things from cards-utils, not add more things to it, so I'd like to see this moved into its own module, `conditions`.

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/ConditionalSectionUtils.java:
```diff
@@ -207,146 +203,66 @@ public static boolean isConditionSatisfied(final Session resourceSession,
-                }
-            }
+            super(values);
+            this.reference = reference;
```
Shouldn't this be part of the default Operand?

marta- committed:
CARDS-2509: Reference question should support conditions based on source form metadata
- Updated the import script to support conditional properties of reference questions

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
- Clean up unused javadoc
- Add missing conditional reference columns to questionnaire import
- Add missing null checks

sashaandjic commented:
Steps to reproduce:
- create **Reference Test - User Answers** form - fill in all fields ( I used `text one`, `text one in section`, `11`, `11.11`) 
- create another **Reference Test - User Answers** form - fill in all fields ( I used `text two`, `text two in section`, `22`, `22.22`) 
- create Reference **Test - Copied Answers** --> observe --> the form has values from the first form and all look good
- go back to the first **Reference Test - User Answers** form and remove all values entered except in the `Text question` field
- open **Reference Test - User Answers** form --> observe --> everything look good
- create another **Reference Test - User Answers** form - fill in all fields ( I used `text three`, `text three in section`, `33`, `33.33`) 
-  open **Reference Test - User Answers** form --> observe --> the form has values from the first and the last form
![testRes](https://github.com/user-attachments/assets/eabade05-d0b9-46ee-b23a-362224a617b5)

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
- Rebase ontop of latest dev to pull in other reference question changes
- Code cleanup:
  - Refactor null checks to remove double negatives
  - Move condition utils to it's own module
  - Fix test form naming issues
  - Clean up constant string naming and reuse utility strings when available

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
Improve conditional references to not change source answers if a new valid source is edited while
  the original source does not meet the condition

acrowthe committed:
CARDS-2509: Reference question should support conditions based on source form metadata
Cleanup: Remove ConditionalOperand and move functionality to base Operand
